cmd/argo/commands/submit.go

```diff
@@ -248,3 +271,100 @@ func waitOrWatch(workflowNames []string, cliSubmitOpts cliSubmitOpts) {
 		watchWorkflow(workflowNames[0])
 	}
 }
+
+func parseParameters(opts *util.SubmitOpts) ([]wfv1.Parameter, error) {
+	newParams := make([]wfv1.Parameter, 0)
+	if len(opts.Parameters) > 0 || opts.ParameterFile != "" {
+		passedParams := make(map[string]bool)
+		for _, paramStr := range opts.Parameters {
+			parts := strings.SplitN(paramStr, "=", 2)
+			if len(parts) == 1 {
+				return nil, fmt.Errorf("Expected parameter of the form: NAME=VALUE. Received: %s", paramStr)
+			}
+			param := wfv1.Parameter{
+				Name:  parts[0],
+				Value: &parts[1],
+			}
+			newParams = append(newParams, param)
+			passedParams[param.Name] = true
+		}
+		if opts.ParameterFile != "" {
+			var body []byte
+			var err error
+			if cmdutil.IsURL(opts.ParameterFile) {
+				body, err = util.ReadFromUrl(opts.ParameterFile)
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				body, err = ioutil.ReadFile(opts.ParameterFile)
+				if err != nil {
+					return nil, err
+				}
+			}
+			yamlParams := make(map[string]string)
+			err = yaml.Unmarshal(body, &yamlParams)
+			if err != nil {
+				return nil, err
+			}
+
+			for k, v := range yamlParams {
+				value, err := strconv.Unquote(string(v))
+				if err != nil {
+					value = string(v)
+				}
+				param := wfv1.Parameter{
+					Name:  k,
+					Value: &value,
+				}
+				if _, ok := passedParams[param.Name]; ok {
+					continue
+				}
+				newParams = append(newParams, param)
+				passedParams[param.Name] = true
+			}
+		}
+	}
+	return newParams, nil
+}
+
+func replaceGlobalParameters(fileContents [][]byte, submitOpts *util.SubmitOpts, cliOpts *cliSubmitOpts) ([][]byte, error) {
+	var output [][]byte
+	for _, body := range fileContents {
+		workflowRaw := make(map[interface{}]interface{})
+		err := yaml.Unmarshal(body, &workflowRaw)
+		if err != nil {
+			return nil, err
+		}
+		spec, ok := workflowRaw["spec"].(map[interface{}]interface{})
+		if !ok {
+			log.Fatalf("Problem with the spec defintion for the workflow: '%s'", workflowRaw["spec"])
+		}
+		args, err := yaml.Marshal(spec["arguments"])
+		if err != nil {
+			return nil, err
+		}
+		var argSpec wfv1.Arguments
+		err = yaml.Unmarshal(args, &argSpec)
+		if err != nil {
+			return nil, err
+		}
+		globalParams := make(map[string]string)
+		for _, param := range argSpec.Parameters {
+			globalParams["workflow.parameters."+param.Name] = *param.Value
+		}
+		newParams, err := parseParameters(submitOpts)
+		if err != nil {
+			return nil, err
+		}
+		globalReplacedTmplStr := string(body)
+		for _, param := range newParams {
+			globalParams["workflow.parameters."+param.Name] = *param.Value
+		}
+		for key, value := range globalParams {
+			globalReplacedTmplStr = strings.ReplaceAll(globalReplacedTmplStr, `"{{`+key+`}}"`, value)
+		}
```

Comment:
This is not how we do string replacement in Argo. We use something called fasttemplates. Find an example in the codebase.

Reply:
I checked out fast template and initial used it. However fast template have the issue that it will substitute a bit more than just the 

> Sorry for the delay @NikeNano. A step in the right direction, but still have more comments.

No worries. Thanks for the feedback, I will update and answer your questions.

Reply:
I initially used fast template but it resulted in that the ``` "``` are removed for the parameters which are not replaced as well meaning:
 ```"{{inputs.parameters.message1}}" ``` -> ```{{inputs.parameters.message1}}```  

if I use 

``` go 
fstTmpl := fasttemplate.New(globalReplacedTmplStr, `"{{`, `}}"`)
```
or 

 ```"{{inputs.parameters.message1}}" ``` -> ```{{inputs.parameters.message1}}```  

if I use 

``` go 
fstTmpl := fasttemplate.New(globalReplacedTmplStr, `"{{`, `}}"`)
```

but then the integers will end up with ```" ```around them as well. My solution to it was to use the string replace instead since I could not find a work around for it. Happy for suggestions :) . 

Here is an example: https://play.golang.org/p/kIlWAu_5unk